src/sdi-progress-dock.c

```diff
@@ -0,0 +1,97 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "sdi-progress-dock.h"
+#include "com.canonical.Unity.LauncherEntry.h"
+#include <glib/gi18n.h>
+#include <snapd-glib/snapd-glib.h>
+
+/**
+ * This class manages the progress bars in the dock for each snap
+ * being updated.
+ */
+
+struct _SdiProgressDock {
+  GObject parent_instance;
+
+  GApplication *application;
+  UnityComCanonicalUnityLauncherEntry *unity_manager;
+};
+
+G_DEFINE_TYPE(SdiProgressDock, sdi_progress_dock, G_TYPE_OBJECT)
+
+/**
+ * This callback should be connected to the `update-progress` signal from a
+ * #sdi_refresh_monitor object. It will receive the total number of tasks and
+ * how many have been done, and if the task has been completed, and with that
+ * will update the progress bars in the dock.
+ */
+void sdi_progress_dock_update_progress(SdiProgressDock *self, gchar *snap_name,
+                                       GStrv desktop_files,
+                                       gchar *task_description,
+                                       guint done_tasks, guint total_tasks,
+                                       gboolean task_done, gpointer data) {
+  if ((desktop_files != NULL) && (total_tasks != 0)) {
+    for (gchar **desktop_file = desktop_files; *desktop_file != NULL;
+         desktop_file++) {
+      // Update dock progress bar
+      g_autoptr(GVariantBuilder) builder =
+          g_variant_builder_new(G_VARIANT_TYPE("a{sv}"));
+      g_variant_builder_add(
+          builder, "{sv}", "progress",
+          g_variant_new_double(((gfloat)done_tasks) / ((gfloat)total_tasks)));
+      g_variant_builder_add(builder, "{sv}", "progress-visible",
+                            g_variant_new_boolean(!task_done));
+      g_variant_builder_add(builder, "{sv}", "updating",
+                            g_variant_new_boolean(!task_done));
+
+      g_autoptr(GVariant) values =
+          g_variant_ref_sink(g_variant_builder_end(builder));
+      unity_com_canonical_unity_launcher_entry_emit_update(
+          self->unity_manager, *desktop_file, values);
+    }
+  }
+}
+
+static void sdi_progress_dock_dispose(GObject *object) {
+  SdiProgressDock *self = SDI_PROGRESS_DOCK(object);
+
+  g_clear_object(&self->unity_manager);
+  g_clear_object(&self->application);
+
+  G_OBJECT_CLASS(sdi_progress_dock_parent_class)->dispose(object);
+}
+
+static void sdi_progress_dock_class_init(SdiProgressDockClass *klass) {
+  GObjectClass *gobject_class = G_OBJECT_CLASS(klass);
+
+  gobject_class->dispose = sdi_progress_dock_dispose;
+}
+
+static void sdi_progress_dock_init(SdiProgressDock *self) {}
+
+SdiProgressDock *sdi_progress_dock_new(GApplication *application) {
+  SdiProgressDock *self = g_object_new(SDI_TYPE_PROGRESS_DOCK, NULL);
+  self->application = g_object_ref(application);
```

Comment:
Isn't this one too many refs?

Reply:
Who owns the other reference? It's leaked at the end of this function so the memory will never be freed

Reply:
The other reference is created in main function, at line 173, which has a g_autoptr() definition, and is passed as the first parameter when `do_startup()` is called.

Reply:
And is passed as a parameter to `sdi_progress_dock_new()` at line 94 in `main.c`.

Reply:
Sorry, i misread the code

Reply:
No problem. That part is a little bit convoluted.